### PR TITLE
Docker, gunicor and gevent

### PR DESCRIPTION
**Pull request by loicgasser**

### CODE REVIEW

File: Dockerfile.mako [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM ubuntu
+MAINTAINER loicgasser
+ENV PROXYWMS_ACCESS_KEY ${proxywms_access_key}
+ENV PROXYWMS_SECRET_KEY ${proxywms_secret_access_key}
+ENV PROXYWMS_BUCKET_NAME ${proxywms_bucket_name}
+COPY . /service_proxywms
+WORKDIR /service_proxywms
+RUN apt-get update -y && apt-get install -y python-pip python-dev git build-essential && pip install virtualenv && make cleanall all && apt-get clean
+EXPOSE 80
+CMD ["/service_proxywms/.venv/bin/gunicorn", "--workers", "4", "-k", "gevent", "--bind", "0.0.0.0:9000", "service_proxywms.wsgi"]
```
Comment (danduk82) [inline Dockerfile.mako]: Ideally the starting paramters for the service should come a) through options to the containter, b) through environment variables provided to the container. Please do not hardcode this kind of setup in the dockerfile

File: Makefile [hunk 2]
```diff
@@ -68,15 +69,27 @@ deploydev:
...
 .PHONY: cleanall
 cleanall: clean
 	rm -rf $(INSTALL_DIR)
 
+Dockerfile.mako:
+Dockerfile: Dockerfile.mako guard-PROXYWMS_ACCESS_KEY guard-PROXYWMS_SECRET_KEY guard-PROXYWMS_BUCKET_NAME
+	${MAKO_CMD} \
+	  --var "proxywms_access_key=$(PROXYWMS_ACCESS_KEY)" \
+	  --var "proxywms_secret_access_key=$(PROXYWMS_SECRET_KEY)" \
+	  --var "proxywms_bucket_name=$(PROXYWMS_BUCKET_NAME)" $< > $@
```
Comment (danduk82) [inline Makefile]: typically these variables should be provided in the runtime environment variables, and not statified in the dockerfile

File: README.md [hunk 3]
```diff
@@ -35,3 +35,9 @@ The steps are:
 
 1. Request the WMS Server image
 2. Return the WMS image to the client
+
+Docker
+------
+
+1. Create the image with `make dockerize`
+2. To run it use `sudo docker run service_proxywms`
```
Comment (danduk82) [inline README.md]: tip: if you add your user to the "docker" group, you might run docker without "sudo"

File: service_proxywms/main.py [hunk 4]
```diff
@@ -150,20 +156,20 @@ def prepare_wmts_response(s3_object):
...
     response.headers.add('Access-Control-Allow-Origin', '*')
     response.headers.add('Access-Control-Allow-Methods', 'GET')
     response.headers.add('Access-Control-Allow-Headers', 'Content-Type, Authorization, x-requested-with, Origin, Accept')
     return response
 
 
-@app.route('/checker')
+@application.route('/checker')
 def checker():
     return 'OK'
```
Comment (danduk82) [inline service_proxywms/main.py]: does this respond ONLY after the whole thing is up? If yes it is a sufficient healthcheck. We want to ensure that the load balancer does not send traffic the container is fully up

Comment (loicgasser) [inline service_proxywms/main.py]: Yes, this responds only after the app is ready.

File: Makefile (hunk 2 above)
Comment (loicgasser) [inline Makefile]: Indeed, but his require a docker compose module right?

Comment (danduk82) [inline Makefile]: We use docker-compose indeed in production. It's one of ranchers base dependencies. It adds a rancher compose on top of docker compose.

Comment (loicgasser) [inline Makefile]: yep, I am looking into it right now. This is where you would tell where to put the logs as well, and create dev, int, prod envs from the main Dockerfile right?

Comment (danduk82) [inline Makefile]: It is what allows you to create "compositions" of services, that work toghether.

**Comment by danduk82** [comment]
Here is the testlink running on rancher :
https://service-proxywms.dev.bgdi.ch

please test ;)

**Comment by loicgasser** [comment]
Thanks @danduk82 ! It's pretty dam good on rancher.
http://codepen.io/loicgasser/full/mmMMZm/

How many workers did you use?

**Comment by danduk82** [comment]
For the moment 4 worker per container, and there are 2 containers. I am not so convinced about the gunicorn setup, I'll put a frontal nginx to handle the connections, and maybe change to uwsgi... 
Moreover, I'd like to test nginx to do the black magic usually implemented in varnish (vanrnish does not support multiple A DNS entries for the backend...).

**Comment by danduk82** [comment]
I clicked on the wrong button, sorrry

**Comment by loicgasser** [comment]
+1 for frontal nginx, my initial idea was to package nginx with it, so use it for sure. Uwsgi switch should be trivial. I d like to see how pypy compares to cpython as well. Maybe we can give it a try. We can probably use a lot more workers though.

**Comment by loicgasser** [comment]
I have optimized the current setup a little. 
According to what I red here:
https://github.com/benoitc/gunicorn/issues/1045
The basline formula for choosing the number of threads and workers is:
CPUs * 2 + 1
Assuming we have 1 or 2 CPUs that means we could handle 5*5 connections simultanously on one instance. 
With 2 instances 50 (16 before) connections. Frontal nginx should be used only to buffer the incoming load, to protect the backend from peaks. ON the other hand I think that we se should completely deactivate the caching at ngnix level, as every request is designed to be sent only a single time to this backend.
I changed the image and pushed a new version so we can try it out. Anyway we ll need to benchmark this and try out uwsgi. I am not sure how to deploy the image to dev yet @danduk82 ;)

**Pull request merged**